File: backend/internal/models/transaction.go
package models

import (
	"time"
)

type Transaction struct {
	ID              uint      `gorm:"primaryKey" json:"id"`
	UserID          uint      `gorm:"not null;index" json:"user_id"`
	CategoryID      *uint     `gorm:"index" json:"category_id"`
	Amount          float64   `gorm:"not null" json:"amount"`
	Type            string    `gorm:"not null;index" json:"type"` // "income" or "expense"
	Description     string    `json:"description"`
	TransactionDate time.Time `gorm:"not null;index" json:"transaction_date"`
	Source          string    `gorm:"default:manual" json:"source"` // "manual", "pdf", "gmail", "invoice"
	CreatedAt       time.Time `json:"created_at"`
	UpdatedAt       time.Time `json:"updated_at"`

	// Associations
	User     User      `gorm:"foreignKey:UserID" json:"user,omitempty"`
	Category *Category `gorm:"foreignKey:CategoryID" json:"category,omitempty"`
}

// TableName specifies the table name
func (Transaction) TableName() string {
	return "transactions"
}

File: frontend/src/types/transaction.ts
export interface Transaction {
  id: number
  user_id: number
  category_id?: number
  amount: number
  type: 'income' | 'expense'
  description: string
  transaction_date: string
  source: string
  tags: string[]
  created_at: string
  updated_at: string
  category?: Category
}

export interface Category {
  id: number
  name: string
  type: 'income' | 'expense'
  icon?: string
  color?: string
  created_at: string
}

export interface CreateTransactionRequest {
  category_id?: number
  amount: number
  type: 'income' | 'expense'
  description: string
  transaction_date: string
  source?: string
  tags?: string[]
}

export interface UpdateTransactionRequest {
  category_id?: number
  amount?: number
  type?: 'income' | 'expense'
  description?: string
  transaction_date?: string
  tags?: string[]
}

export interface TransactionFilter {
  type?: 'income' | 'expense'
  start_date?: string
  end_date?: string
  category_id?: number
  q?: string
  tags?: string
  min_amount?: number
  max_amount?: number
  page?: number
  page_size?: number
}

export interface TransactionListResponse {
  data: Transaction[]
  total: number
  page: number
  page_size: number
}
